cedar-policy-core/src/authorizer/partial_response.rs

```diff
@@ -0,0 +1,615 @@
+        false_permits: impl IntoIterator<Item = (PolicyID, (ErrorState, Arc<Annotations>))>,
+        residual_permits: impl IntoIterator<Item = (PolicyID, (Arc<Expr>, Arc<Annotations>))>,
+        true_forbids: impl IntoIterator<Item = (PolicyID, Arc<Annotations>)>,
+        false_forbids: impl IntoIterator<Item = (PolicyID, (ErrorState, Arc<Annotations>))>,
+        residual_forbids: impl IntoIterator<Item = (PolicyID, (Arc<Expr>, Arc<Annotations>))>,
+        errors: impl IntoIterator<Item = AuthorizationError>,
+    ) -> Self {
+        Self {
+            satisfied_permits: true_permits.into_iter().collect(),
+            false_permits: false_permits.into_iter().collect(),
+            residual_permits: residual_permits.into_iter().collect(),
+            satisfied_forbids: true_forbids.into_iter().collect(),
+            false_forbids: false_forbids.into_iter().collect(),
+            residual_forbids: residual_forbids.into_iter().collect(),
+            errors: errors.into_iter().collect(),
+            true_expr: Arc::new(Expr::val(true)),
+            false_expr: Arc::new(Expr::val(false)),
+        }
+    }
+
+    /// Convert this response into a concrete evaluation response.
+    /// All residuals are treated as errors
+    pub fn concretize(self) -> Response {
+        self.into()
+    }
+
+    /// Attempt to reach a partial decision; the presence of residuals may result in returning [`None`],
+    /// indicating that a decision could not be reached given the unknowns
+    pub fn decision(&self) -> Option<Decision> {
+        match (
+            !self.satisfied_permits.is_empty(),
+            !self.false_permits.is_empty(),
+            !self.residual_permits.is_empty(),
+            !self.satisfied_forbids.is_empty(),
+            !self.false_forbids.is_empty(),
+            !self.residual_forbids.is_empty(),
+        ) {
+            // Any true forbids means we will deny
+            (_, _, _, true, _, _) => Some(Decision::Deny),
+            // No potentially or trivially true permits, means we default deny
+            (false, _, false, _, _, _) => Some(Decision::Deny),
+            // Potentially true forbids, means we don't know
+            (_, _, _, _, _, true) => None,
+            // No true permits, but some potentially true permits + no true/potentially true forbids means we don't know
+            (false, _, true, false, _, false) => None,
+            // At least one trivially true permit, and no trivially or possible true forbids, means we allow
+            (true, _, _, false, _, false) => Some(Decision::Allow),
+        }
+    }
+
+    /// All of the [`Effect::Permit`] policies that were known to be satisfied
+    fn satisfied_permit_ids(&self) -> impl Iterator<Item = &PolicyID> {
+        self.satisfied_permits.iter().map(first)
+    }
+
+    /// All of the [`Effect::Forbid`] policies that were known to be satisfied
+    fn satisfied_forbid_ids(&self) -> impl Iterator<Item = &PolicyID> {
```

Comment:
for these two, for symmetry with `definitely_satisfied()`, we should either rename these `satisfied_permits()` and `satisfied_forbids()`, or rename `definitely_satisfied()` to `definitely_satisfied_ids()`.  I'm also unsure whether we should add the word "definitely" to these two methods' names